File: algo_05_dynamic_plan/dp_2_subsequence/L300.cpp


//leetcode 300 最长上升子序列问题
//DP时间复杂度O(n*n)

#include <iostream>
#include <vector>

using namespace std ;
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
        int maxL = 1;
        //memo[i] 表示 以nums[i] 结尾的最长上升子序列的长度
        vector<int> memo(nums.size(), 1) ;
        for(int i = 1 ; i < nums.size() ; i ++){
            for(int j = 0 ; j < i ; j ++){
                if(nums[i] > nums[j])
                    memo[i] = max(memo[i] , 1 + memo[j]);
            }
            maxL = max(maxL,memo[i]);
        }
        return maxL;
    }
};

int main(){
    //int arry[] = {1, 7 , 3 , 5 , 9 , 4 , 8};
    int arry[] = {10,9,2,5,3,7,101,18};
    vector<int> nums(arry ,  arry + 7);
    cout << Solution().lengthOfLIS(nums) ;
}

// 做题体会！！！
// 有一类问题的写法，包括双指针，dp,滑动窗口，都是以当前遍历到的元素i作为一个考察点，
// 然后从0向后遍历到i或者从i往前倒着走来解决问题


//这道题还有一种nlogn的解法，还没做,不太好理解贪心+二分查找

File: python-algorithm/Big_Company_Recent/bytedance/L300.py
# -*- coding: utf-8 -*-

# 动态规划O(n*n)
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        sz = len(nums)
        dp = [1]*sz  # dp[i]表示以nums[i]作为结尾的最长上升子序列
        max_len = 1
        for i in range(1, sz):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
            max_len = max(max_len, dp[i])
        return max_len


# 二分法；很难想到
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        '''
        (1)新建数组，record[i]最终表示的是长度为i+1的递增子序列中尾数最小的值；因此record是递增的
        (2)在遍历时通过二分法确定元素应该被放到record中的具体位置，因此时间复杂度是N*logN
        '''
        maxL = 0
        record = [0]*(len(nums))
        for nu in nums:
            lt, rt = 0, maxL
            while lt<rt:
                mid = lt + (rt-lt)//2
                if record[mid]<nu:
                    lt = mid + 1
                else:
                    rt = mid
            record[lt] = nu
            if lt == maxL:
                maxL += 1
        return maxL